keps/sig-instrumentation/20191115-structured-logging.md

```diff
@@ -0,0 +1,556 @@
+
+#### Json format
+
+We would like to propose a JSON as a future default format in Kubernetes for both production and development use. Some pros of using JSON:
+* Broadly adopted by logging libraries with very efficient implementations (zap, zerolog).
+* Out of the box support by many logging backends (Elasticsearch, Stackdriver, BigQuery, Splunk)
+* Easily parsable and transformable
+* Existing tools for ad-hoc analysis (jq)
+
+Example:
+
+```json
+{
+   "ts":"00:15:15.525108",
+   "msg":"Updating ready status",
+   "core/node":{
+      "name":"node-1"
+   },
+   "core/pod":{
+      "name":"nginx-1",
+      "namespace":"default"
+   },
+   "probe":{
+      "ready":false
+   }
+}
+```
+
+### Logging configuration
+
+We would like to introduce new logging configuration options shared by all kubernetes components. `LoggingConfig` structure should be implemented as part of `k8s.io/component-base` options and include common set of flags for logger initialization.
+* `--logging-format` flag should allow to pick between logging interface implementation and allow for backward compatibility. Setting this flag will select particular logger implementation from registry.
+* `-v` flag should allow to change verbosity of logging
+
+Any additional configuration needed by loggers should be provided by flags specific to its implementation. Default loggers should provide a basic customization, but it will not be encouraged to avoid feature bloat. Adding new loggers, changing logger behavior or extending its functionality should be done outside of Kubernetes by implementing logr API and registering it in code.
+
+We acknowledge that extension via code is not the best, but should be sufficient for now.
+
+Proposed flag `--logging-format`  values:
+* `legacy` for old logging format
+* `json` for json format
+
+All logs from the `json` logger should go to standard output.  Out of the existing klog command line arguments only ‘-v’ with be supported in future.
+
+### Kubernetes log schema
+
+Introducing structured logging API is not enough to address one of the biggest issues with log creation or analysis, that is inconsistent metadata. Size of Kubernetes codebase could easily result in drifting naming conventions. As part of this proposal we want to propose a mechanism for ensuring consistency of metadata. To achieve this we will introduce a way to define a schema for subset of commonly used fields in kubernetes. This schema should improve reusability of fields and provide discovery mechanism for logging backend vendors.
+
+Introducing this schema will include three things:
+* Schema definition format
+* Set of helper functions to populate this schema (e.g. `WithObject` for referencing k8s object)
```

Comment:
objects are logically singular.  Many (most?) control loops have inputs from multiple sources that are all important for debugging a particular outcome.  An exploration of how that will work is important.

Reply:
That situation is extremely common.  Every update, patch, apply operator in the apiserver and control loops managing ClusterRoles come to mind immediately.

Reply:
Ok, if I get it right I think we have two a bit separate cases here:
1) we have two different Pods which are processed in some way in a single request or pass of a controller,
2) we have a single Pod for which the state is changing during the request handling.
If this is correct then from the perspective of structured logs the second case is not that interesting because we only want to log data which identifies given resource like name and namespace and not the whole actual state of this resource so context does not need to contain the actual resource object but only its identification.

Reply:
I presume you could `WithObject` any number of objects, but this should clarify that and what happens if you have 2 of the same kind (since it seems to propose Kind as the key)